### PR TITLE
Sp3 store recycle bin locations on firebase

### PR DESCRIPTION
**Pull request by LucaCarroz**
This PR adds the following functionnalities to the object `FirebaseDB`:

-  `storeBinLocation` can be called to store the location of a bin as a pair of longitude and latitude. We're using the type `LatLng` since this is the type of a pin on Google Maps.
- `getAllBinLocations` can be called to get the list of all pins that we can then display on the map, to let the user know where the nearest recycling bin is. We will likely add more informations later, by example to diferentiate recycle bin from trash cans, etc.

### CODE REVIEW

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt [hunk 1]
```diff
@@ -293,4 +301,40 @@ object FirebaseDB {
...
+        // Looks like manually deleting the unchecked cast is the only way to remove the warnings
+        @Suppress("UNCHECKED_CAST")
+        val childrens = databaseReference.child(BIN_LOCATION_PATH).get().await().value
+                as? Map<String, Any> ?: return null
+
+        val locations = mutableListOf<LatLng>()
+        for (child in childrens.values) {
+            // Converts the Map back to a LatLng. If some child does not have the correct format, skip it
+            @Suppress("UNCHECKED_CAST")
+            val location = child as? HashMap<String, Double>? ?: child as? HashMap<String, Long>? ?: continue
```
Comment (iliasmerigh) [inline app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:333]: This way of iterating over element with the `continue` operator can be hard to understand. I recommend you to keep an easier code to understand like that :

```
for (child in childrens.values) {
    // Converts the Map back to a LatLng. If some child does not have the correct format, skip it
    @Suppress("UNCHECKED_CAST")
    val location = child as? HashMap<String, *>?
    if (location != null && location.containsKey("latitude") && location.containsKey("longitude")) {
        val lat = location["latitude"] as? Double
        val lng = location["longitude"] as? Double
        if (lat != null && lng != null) {
            locations.add(LatLng(lat, lng))
        }
    }
}

```

But feel free to use the one you are more confortable with. 

Moreover, I don't understand why coordinates could be `Double` or `Long` ? Shouldn't we fix to one unique format ?

**Review by iliasmerigh — CHANGES_REQUESTED** [review]
Great work on this PR ! The refactoring makes the code easier to understand, and the "recycle bin" methods will be useful for me very soon.

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt [hunk 2]
```diff
@@ -293,4 +301,40 @@ object FirebaseDB {
...
     }
+
+    /**
+     * Store the given [location] as the location of a recycling bin.
+     *
+     * @param location the location of the bin
+     *
+     * @return true if the location got stored in the database, and false if it failed
+     */
+    suspend fun storeBinLocation(location: LatLng): Boolean{
```
[the remark this answers is not in the document]
Comment (LucaCarroz) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:312]: Great ideas! I actually thought about adding up and down votes, such that a bin gets removed if it has a too low ratio of up to down votes (to prevent some mailcious user from deleting every markers). I think we can add this feature once google maps is mergde into the app.

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt [hunk 3]
```diff
@@ -293,4 +301,40 @@ object FirebaseDB {
...
+        databaseReference.child(BIN_LOCATION_PATH).child(freshId).setValue(location).await()
+        return true
+    }
+
+    /**
+     * Retrieves the list of all bin locations currently present in the database
+     *
+     * @return the list of locations of all the bins, or null if an error happened
+     */
+    suspend fun getAllBinLocations(): List<LatLng>?{
```
[the remark this answers is not in the document]
Comment (LucaCarroz) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:323]: The `Bin` class is a great idea indeed! As you said tho, we better wait for the implementation of Google maps to be in place to know exactly how we want to define the class. 
The conversion from `List` to `Set` is a great idea, I will apply this, thanks!

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt (hunk 1 above)
Comment (LucaCarroz) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:333]: I personaly prefer the `continue` to explicitly show when we skip and element, and using Kotlins `?:` which are quite convenient here.
Also, it looks like if we store a LatLng with integer values (by example `LatLng(1.0, 2.0)`), firebase will store it as 2 `Long` instead of 2 `Double`. Eventhough this should rarely happen, since a latitude or longitude is way more likely to have decimals, this second cast handles this corner case.

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/firebase/FirebaseDBTest.kt [hunk 4]
```diff
@@ -159,4 +167,42 @@ class FirebaseDBTest {
...
+    }
+
+    @Test
+    fun storeBinLocationCorrectlyUpdatesDatabase() {
+
+        val binLocation = LatLng(12.3, -43.0)
+
+        runBlocking {
+
+            assertTrue(FirebaseDB.storeBinLocation(binLocation))
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/firebase/FirebaseDBTest.kt:184]: No it's fine. I said that only because a team member suggested me to do like that but if the project already contains both of them there is not problem

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:312]: Good idea. I will take it into consideration during my work.

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt (hunk 1 above)
Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:333]: I see... Anyway, the `Bin` class that I will create will cast it to a float. For next sprint, please let me now if you find a solution to force the cast on firebase side.

**Review by iliasmerigh — APPROVED** [review]
Thank you for applying the requested changes, and for your explanations in the comments.
